server/src/main/java/io/littlehorse/common/model/getable/core/taskrun/TaskRunModel.java

```diff
@@ -298,21 +336,94 @@ public void updateTaskResult(ReportTaskRunModel ce) {
             // Tell the WfRun that the TaskRun is done.
             transitionTo(TaskStatus.TASK_SUCCESS);
         } else if (shouldRetry()) {
-            transitionTo(TaskStatus.TASK_SCHEDULED);
-            scheduleAttempt();
+            scheduleRetryAtAppropriateTime();
         } else {
             transitionTo(ce.getStatus());
         }
 
+        // The WfRun may need to advance.
         processorContext.getableManager().get(getWfRunId()).advance(ce.getTime());
     }
 
     /**
-     * Returns the ID of the WfRun that this TaskRun belongs to.
-     * @return the ID of the WfRun that this TaskRun belongs to.
+     * Called when a TASK_PENDING retry attempt is now ready to launch according to the retry backoff
+     * policy.
      */
-    public WfRunIdModel getWfRunId() {
-        return taskRunSource.getSubSource().getWfRunId();
+    public void markAttemptReadyToSchedule() {
+        dispatchTaskToQueue();
+    }
+
+    /**
+     * Dispatches the latest TaskAttempt to the server's Task Queues.
+     */
+    public void dispatchTaskToQueue() {
+        if (attempts.isEmpty()) {
+            throw new IllegalStateException("can't have empty attempts queue");
+        }
+        TaskAttemptModel attempt = attempts.get(attempts.size() - 1);
+        if (attempt.getStatus() != TaskStatus.TASK_PENDING) {
+            throw new IllegalStateException("Cannot schedule task attempt that isn't in PENDING state");
+        }
+
+        attempt.setStatus(TaskStatus.TASK_SCHEDULED);
+        attempt.setScheduleTime(new Date());
+
+        ScheduledTaskModel scheduledTask = new ScheduledTaskModel();
+        scheduledTask.setVariables(inputVariables);
+        scheduledTask.setAttemptNumber(attempts.size() - 1);
+        scheduledTask.setCreatedAt(new Date());
+        scheduledTask.setSource(taskRunSource);
+        scheduledTask.setTaskDefId(taskDefId);
+        scheduledTask.setTaskRunId(id);
+
+        processorContext.getTaskManager().scheduleTask(scheduledTask);
+    }
+
+    private boolean shouldRetry() {
+        // Shouldn't look at nodeRun to decide whether to retry task or not.
+        TaskAttemptModel latest = getLatestAttempt();
+        if (latest == null) {
+            // This really shouldn't happen I think
+            return false;
+        }
+
+        if (latest.getStatus() != TaskStatus.TASK_FAILED && latest.getStatus() != TaskStatus.TASK_TIMEOUT) {
```

Comment:
[Suggestion / Feel free to ignore]:
new method in the `TaskAttemptModel#shouldRetry`  that contains this line here